oauthenticator/oauth2.py

```diff
@@ -316,6 +328,41 @@ class OAuthenticator(Authenticator):
         """,
     )
 
+    allowed_groups = Set(
+        Unicode(),
+        config=True,
+        help="""
+        Allow members of selected groups to log in.
+        """,
+    )
+
+    admin_groups = Set(
+        Unicode(),
+        config=True,
+        help="""
+        Allow members of selected groups to sign in and consider them as
+        JupyterHub admins.
+
+        If this is set and a user isn't part of one of these groups or listed in
+        `admin_users`, a user signing in will have their admin status revoked.
+        """,
+    )
+
+    auth_state_groups_key = Union(
+        [Unicode(), Callable()],
+        config=True,
+        help="""
+        Determine groups this user belongs based on contents of auth_state.
+
+        Can be a string key name (use periods for nested keys), or a callable
+        that accepts the auth state (as a dict) and returns the groups list.
+
+        This configures how group membership in the upstream provider is determined
```

Comment:
Does `upstream provider` mean JupyterHub?

Reply:
@manics I copied this from the current docstring (https://github.com/jupyterhub/oauthenticator/blob/79db03cb901da09301e942daeebd0f9b6f383446/oauthenticator/generic.py#L30) but I agree it's actually quite confusing. I'll think about it and try to rewrite it.

Reply:
@manics I spent some time rewording this, and this led me to the conclusion that we should just not allow usage of these without `manage_groups` also being `True`. I've added a more full rationale under 'Backwards compatibility' describing it. I've also rewritten the docstring now! Let me know what you think :)